clang/docs/HIPSupport.rst

```diff
@@ -176,3 +176,34 @@ Predefined Macros
    * - ``HIP_API_PER_THREAD_DEFAULT_STREAM``
      - Alias to ``__HIP_API_PER_THREAD_DEFAULT_STREAM__``. Deprecated.
 
+C++20 Concepts with HIP and CUDA
+--------------------------------
+
+In Clang, when working with HIP or CUDA, it's important to note that all constraints in C++20 concepts are assumed to be for the host side only. This behavior is consistent across both programming models, and developers should be aware of this assumption when writing code that utilizes C++20 concepts.
+
+Example:
+.. code-block:: c++
+
+   template <class T>
+   concept MyConcept = requires(T& obj) {
+     my_function(obj);  // Assumed to be a host-side requirement
```

Comment:
How would one write a correct GPU-side requirement?

E.g. I want some `__device__` function to be callable.

Reply:
currently, we do not support that. 

I would suggest adding an extension to the clang that allows `__host__` and `__device__` attributes on call expressions in concept definition to indicate the required callability for the host or device.

For example,

```
template <class T>
   concept MyConcept = requires(T& obj) {
     __device__ my_function(obj);  // requires my_function(obj) callable on device side
}
```

Reply:
If I understand correctly, normally a template is usable from either host or device (depending on whether it ends up calling any host-only or device-only function). This choice for concepts seems like it's going to be problematic for that model. Something as simple as:

```c++
template<Copyable T> T f(T x) { return x; }
```

... should really be callable on the host or device side if `T` is copyable on the host or device side, and using the host side in all cases will mean that things like the C++ `<complex>` or `<functional>` header may stop doing the right thing in some cases if/when they get extended to use concepts. And it seems like with this patch there's not anything that the authors of those headers can really do about it.

Perhaps it would be better for the host/device choice in a concept satisfaction check to depend on the context in which the concept is required to be satisfied (which I would imagine is what happened by chance before this patch), and for us to include the CUDA context as part of the constraint satisfaction cache key? That kind of direction seems like it'd give closer results to what we'd get from the split compilation model. I don't know if that actually works in general, though. For example, given:

```c++
__host__ X<T> host_global;
__device__ X<T> device_global;
```

... where `X` is a constrained template, what seems like it should happen here is that we take the `__host__` / `__device__` into account when concept-checking `X`'s template arguments, but I'd worry that we don't have the host/device information to hand when checking the concept satisfaction query for `X`.

More broadly, I think there'll be cases where a CUDA developer will want, from host code, to check whether a constraint would be satisfied on the device, and some mechanism for doing that seems useful. I think that *can* be done with the model I suggest above, by putting a kernel call inside a `requires` expression, but it seems awkward, so perhaps some syntax for explicitly evaluating a *concept-id* in a particular host/device would be useful.

But it definitely seems worthwhile to figure out what rule NVCC is using here.

Reply:
> If I understand correctly, normally a template is usable from either host or device (depending on whether it ends up calling any host-only or device-only function). 

No. template function in CUDA/HIP is subject to the same availability check regarding host/device attributes. e.g https://godbolt.org/z/ccTxhEhfo

constexpr template functions can be called by both host and device functions not because they are template functions, but because host device attributes are added to them implicitly. They are still subject to host/device availability checks.

Reply:
For a typical use case of concept in CUDA programs, please see https://godbolt.org/z/o7Wa68n9c

This is taken from issue https://github.com/llvm/llvm-project/issues/67507.

In this example, users want to express two constraints on geometric_shape:

1. it can be passed to a function draw

2. it can be passed to a function area and the result is convertible to double

For the first constraint, users only need it on the host side. For the second constraint, users need it on both the host side and the device side. This gives us some insight into users' needs for constraints: they are usually different for host and device sides, since users may want to do different things on host and device sides. Therefore, assuming a constraint in a concept should be satisfied on both the device and host sides will result in some unnecessary extra constraints on either side.

Is it OK to evaluate the constraints by the context where the template is instantiated? For example, when we instantiate the kernel `template <geometric_shape T> __global__ void compute_areas`, can we evaluate the constraints in the device context to get what we need? It is not good. Because then the constraint about function draw needs to be satisfied on the device side. That is not what we need. The point is, that the constraints defined in a concept need to have individual required context. We want to be able to express that this constraint should be satisfied in the device context, and that constraint should be satisfied in the host context. That is why I propose to allow `__device__` and `__host__` attributes to be added to the call expressions in concepts to indicate the required context for an individual constraint.

Now that we have discussed the users' needs regarding device/host contexts of constraints. Let's look at how nvcc currently evaluates satisfaction of constraints.

Based on https://godbolt.org/z/o7Wa68n9c , the instantiation of `work<triangle>` is successful. We can infer that `triangle` satisfies the two constraints. They can only be evaluated in the host context since functions `draw` and `area` are all host functions. Even though the instantiation of `work<triangle>` is done in a device context, the evaluation of the constraints is still done in the host context.

The current patch matches nvcc's behaviour.